src/wallet/test/wallet_transaction_tests.cpp

```diff
@@ -4,12 +4,187 @@
+        const CWalletTx* multisig_tx = SendCoinsAndGenBlock(*this, GetScriptForDestination(WitnessV0ScriptHash(multisig_script)), 3 * COIN, *wallet, coinbase_dest_script);
+
+        // Now create a transaction that spends the funds locked in the multisig script
+        CCoinControl coin_control;
+        coin_control.m_allow_other_inputs = false;
+        coin_control.Select({multisig_tx->GetHash(), 0});
+        auto op_spend_tx = Assert(CreateTransaction(*wallet, {{dest_script, 2 * COIN, true}}, 1, coin_control, /*sign=*/false));
+
+        // Now the real test begins, create unsigned tx
+        CMutableTransaction unsigned_tx(*op_spend_tx->tx);
+        std::map<COutPoint, Coin> coins;
+        coins[unsigned_tx.vin[0].prevout] = Coin(multisig_tx->tx->vout[unsigned_tx.vin[0].prevout.n], /*nHeightIn=*/113, /*fCoinBaseIn=*/false);
+
+        // 1) Use the first three keys and add tx to wallet via mempool acceptance:
+        FillableSigningProvider keystore_1;
+        for (int i=0; i<3; i++) BOOST_ASSERT(keystore_1.AddKey(priv_keys[i]));
+        BOOST_ASSERT(keystore_1.AddCScript(multisig_script));
+        std::map<int, bilingual_str> input_errors;
+        CMutableTransaction signed_tx_1 = unsigned_tx;
+        Assert(SignTransaction(signed_tx_1, &keystore_1, coins, SIGHASH_DEFAULT, input_errors));
+
+        const CTransactionRef& spend_from_multisig_tx_1 = MakeTransactionRef(signed_tx_1);
+        wallet->transactionAddedToMempool(spend_from_multisig_tx_1, /*mempool_sequence=*/0);
+        const CWalletTx* wtx_spending_multisig_1 = Assert(WITH_LOCK(wallet->cs_wallet, return wallet->GetWalletTx(signed_tx_1.GetHash())));
+        BOOST_CHECK(wtx_spending_multisig_1->GetWitnessHash() == spend_from_multisig_tx_1->GetWitnessHash());
+
+        // 2) Use the last keys and try to add tx to wallet via block connection:
+        FillableSigningProvider keystore_2;
+        for (int i=priv_keys.size()-1; i>=2; i--) keystore_2.AddKey(priv_keys[i]);
+        BOOST_ASSERT(keystore_2.AddCScript(multisig_script));
+        CMutableTransaction signed_tx_2 = unsigned_tx;
+        Assert(SignTransaction(signed_tx_2, &keystore_2, coins, SIGHASH_DEFAULT, input_errors));
+
+        // Assert that the tx id is equal to the other tx but the witness data is different
+        BOOST_ASSERT(signed_tx_1.GetHash() == signed_tx_2.GetHash());
+        BOOST_ASSERT(signed_tx_1.vin[0].scriptWitness.stack != signed_tx_2.vin[0].scriptWitness.stack);
+
+        // Now connect the block and verify current behavior
+        const CWalletTx* wtx = GenBlockAndRetrieveMinedTx(*this, signed_tx_2, coinbase_dest_script, *wallet);
+
+        // Important: current wallet behavior will NOT update the input witness data.
+        // The wallet will only see the witness data of the first seen tx.
+        BOOST_CHECK(wtx->GetHash() == signed_tx_2.GetHash());
+        BOOST_CHECK(wtx->GetWitnessHash() != MakeTransactionRef(signed_tx_2)->GetWitnessHash());
+    }
```

Comment:
In the other test case with the two p2wpkh transactions, there is a step at the end to reload the wallet from disk. Is there any value in doing the same here?

```
        // Reload the wallet as it would be reloaded from disk and check that the witness data from the first transaction is still there.
        // (flush the previous wallet first)
        wallet->Flush();
        DatabaseOptions options;
        std::unique_ptr<CWallet> wallet_reloaded = std::make_unique<CWallet>(m_node.chain.get(), "", m_args,
                                                                             DuplicateMockDatabase(wallet->GetDatabase(),options));
        BOOST_ASSERT(wallet_reloaded->LoadWallet() == DBErrors::LOAD_OK);
        const CWalletTx* reloaded_wtx_with_original_witness = WITH_LOCK(wallet_reloaded->cs_wallet, return wallet_reloaded->GetWalletTx(wtx->GetHash()));
        BOOST_CHECK_EQUAL(reloaded_wtx_with_original_witness->GetWitnessHash(), MakeTransactionRef(signed_tx_1)->GetWitnessHash());

```

Reply:
Nop, there is no value doing that here.
The test here validates the current limitation of not updating the existent transaction. As there is no wtx update, there is no write to disk, there by there is no reason to test a reload.

And even if the wallet limitation wouldn't exist, the first reload test case is enough to verify the record durability. Sources flow wise, replace a tx with no witness is exactly the same as replacing a tx with witness.